File: maths/nombres/automorphe/automorphe.py
# -----------------------------------------------
# Fonction automorphe
# -----------------------------------------------
# Un palindrome, c'est une chaîne qui de gauche à
# droite, et de droite à gauche possède les memes
# caractères.
# -----------------------------------------------


# -----------------------------------------------
# Version courte
# -----------------------------------------------
# Cette version permet de comprendre le contexte
# de la fonction et sa logique.


def automorphe(nombre):
    longueur = len(str(nombre))

    carre = nombre ** 2

    carre_str = str(carre)
    carre_fin = carre_str[-longueur:]
    carre_fin_nombre = int(carre_fin)

    return carre_fin_nombre == nombre



# -----------------------------------------------
# Application
# -----------------------------------------------

nombre = 76
res = automorphe(nombre)
chaine = ["n'est pas", "est"][res]
print(f"{nombre} {chaine} automorphe")



# -----------------------------------------------
# Version golf
# -----------------------------------------------
# Version condensée et optimisée du code, utili-
# sant des noms de variables courts et combinant
# certaines opérations pour réduire la taille du
# code. Pour la beauté du geste !

A=lambda n:str(n**2)[-len(str(n)):]==str(n)



# -----------------------------------------------
# Version commentee
# -----------------------------------------------
# Proche de la version raccourcie, mais avec des
# commentaires concis pour expliquer les etapes
# principales de la fonction.

def automorphe(nombre):
    # Récupérer la longueur du nombre.
    longueur = len(str(nombre))

    # Elever le nombre au carré.
    carre = nombre ** 2

    # Convertir en string.
    carre_str = str(carre)

    # Récupérer la fin du carré.
    carre_fin = carre_str[-longueur:]

    # Convertir la fin du carré en nombre.
    carre_fin_nombre = int(carre_fin)

    # Retourner la solution.
    return carre_fin_nombre == nombre









#

File: nombre/automorphe/automorphe.js
//-----------------------------------------------
// Fonction automorphe
/*-----------------------------------------------
/ Un nombre automorphe est un entier naturel dont
/ la suite des chiffres de son carré se termine
/ par celle du nombre lui-même.
//---------------------------------------------*/


//-----------------------------------------------
// Version courte
//-----------------------------------------------

function automorphe(nombre) {
    let longueur = nombre.toString().length;

    let carre = Math.pow(nombre, 2)
                    .toString()
                    .slice(-longueur);
    carre = parseInt(carre);

    return carre == nombre;
}



//-----------------------------------------------
// Application
//-----------------------------------------------

let nombres = [125, 76, 89, 625];
for (let nb of nombres) {
    let res = automorphe(nb);
    if (res) {
        console.log(nb, "est automorphe.");
    }
    else {
        console.log(nb, "n'est pas automorphe.");
    }
}



//-----------------------------------------------
// Version golf
//-----------------------------------------------

let A=n=>(c=n**2,c.toString().endsWith(n));


//-----------------------------------------------
// Version detaillee
//-----------------------------------------------

function automorphe_detaillee(nombre) {
    // Récupérer la longueur du nombre.
    let longueur = nombre.toString().length;

    let carre = nombre ** 2;

    let carre_str = carre.toString();
    let carre_fin = carre_str.slice(-longueur);
    let carre_fin_nombre = parseInt(carre_fin);

    return carre_fin_nombre == nombre;
}



//-----------------------------------------------
// Version commentee
//-----------------------------------------------

function automorphe_commentee(nombre) {
    // Récupérer la longueur du nombre.
    let longueur = nombre.toString().length;

    // Elever le nombre au carré.
    let carre = nombre ** 2;

    // Convertir en string.
    let carre_str = carre.toString();

    // Récupérer la fin du carré.
    let carre_fin = carre_str.slice(-longueur);

    // Convertir la fin du carré en nombre.
    let carre_fin_nombre = parseInt(carre_fin);

    // Retourner la solution.
    return carre_fin_nombre == nombre;
}









//